mongorestore/mongorestore_txn_test.go

```diff
@@ -0,0 +1,171 @@
+package mongorestore
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/mongodb/mongo-tools-common/bsonutil"
+	"github.com/mongodb/mongo-tools-common/db"
+	"github.com/mongodb/mongo-tools-common/testtype"
+	"github.com/mongodb/mongo-tools-common/testutil"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+
+	. "github.com/smartystreets/goconvey/convey"
+)
+
+const txnTestDataFile = "testdata/transactions.json"
+
+type txnTestDataMap map[string]*txnTestDataCase
+
+type txnTestDataCase struct {
+	Ops       []db.Oplog `bson:"ops"`
+	NS        string     `bson:"ns"`
+	PostImage []bson.D   `bson:"postimage"`
+}
+
+func TestMongorestoreTxns(t *testing.T) {
+	testtype.SkipUnlessTestType(t, testtype.IntegrationTestType)
+	client, err := testutil.GetBareSession()
+	if err != nil {
+		t.Fatalf("No server available")
+	}
+
+	data, err := readTxnTestData(txnTestDataFile)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Create test collections (if they don't exist) and clear documents.
+	for _, v := range data {
+		parts := strings.SplitN(v.NS, ".", 2)
+		db := client.Database(parts[0])
+		coll := db.Collection(parts[1])
+		err := coll.Drop(context.Background())
+		if err != nil {
+			t.Fatal(err)
+		}
+		res := db.RunCommand(context.Background(), bson.D{{"create", parts[1]}})
+		if res.Err() != nil {
+			t.Fatal(res.Err())
+		}
+	}
+
+	// Create a dump directory from transactions.json
+	dumpPath := createTxnTestDataDir(t, data)
+
+	Convey("With a test MongoRestore", t, func() {
+		args := []string{
+			OplogReplayOption,
+			DropOption,
+			dumpPath,
+		}
+		restore, err := getRestoreWithArgs(args...)
+
+		So(err, ShouldBeNil)
+
+		result := restore.Restore()
+		So(result.Err, ShouldBeNil)
+
+		for k, v := range data {
+			Println("postImageCheck for", k)
+			So(postImageCheck(client, v), ShouldBeNil)
+		}
+	})
+}
+
+func createTxnTestDataDir(t *testing.T, data txnTestDataMap) string {
```

Comment:
just to make sure I understand: this is combining all of the oplog entries from the different test cases in the JSON file so they are applied in some interleaved order, right? (so basically all the tests are happening at once)

Reply:
Yes, added a comment to that effect.